File: api/src/Sudoku/Generator.php
<?php

namespace App\Sudoku;

class Generator
{
    public function __construct(private Solver $solver) {}

    public function generate(): Board
    {
        $board = Board::empty();
        $this->fillDiagonalBoxes($board);
        $this->solver->solve($board);
        $this->removeNumbers($board);
        return $board;
    }

    private function fillDiagonalBoxes(Board $board): void
    {
        // The 3 diagonal 3×3 boxes don't share rows or columns with each other,
        // so they can be filled with random shuffles without any conflict checks.
        for ($box = 0; $box < 3; $box++) {
            $numbers = range(1, 9);
            shuffle($numbers);
            $i = 0;
            for ($r = $box * 3; $r < $box * 3 + 3; $r++) {
                for ($c = $box * 3; $c < $box * 3 + 3; $c++) {
                    $board->setValue($r, $c, $numbers[$i++]);
                }
            }
        }
    }

    private function removeNumbers(Board $board, int $attempts = 60): void
    {
        while ($attempts > 0) {
            $row = rand(0, 8);
            $col = rand(0, 8);
            if ($board->getValue($row, $col) !== 0) {
                $backup = $board->getValue($row, $col);
                $board->setValue($row, $col, 0);
                $copy = Board::fromArray($board->toArray());
                if (!$this->solver->solve($copy)) {
                    $board->setValue($row, $col, $backup);
                }
                $attempts--;
            }
        }
        for ($r = 0; $r < 9; $r++) {
            for ($c = 0; $c < 9; $c++) {
                $board->setReadOnly($r, $c, $board->getValue($r, $c) !== 0);
            }
        }
    }
}

File: api_py/sudoku/generator.py
from __future__ import annotations

import random

from .board import Board
from .solver import Solver


class Generator:
    def __init__(self) -> None:
        self._solver = Solver()

    def generate(self) -> Board:
        board = Board.empty()
        self._fill_diagonal_boxes(board)
        self._solver.solve(board)
        self._remove_numbers(board)
        return board

    def _fill_diagonal_boxes(self, board: Board) -> None:
        for box in range(3):
            numbers = list(range(1, 10))
            random.shuffle(numbers)
            i = 0
            for r in range(box * 3, box * 3 + 3):
                for c in range(box * 3, box * 3 + 3):
                    board.set_value(r, c, numbers[i])
                    i += 1

    def _remove_numbers(self, board: Board, attempts: int = 60) -> None:
        while attempts > 0:
            row = random.randint(0, 8)
            col = random.randint(0, 8)
            if board.get_value(row, col) != 0:
                backup = board.get_value(row, col)
                board.set_value(row, col, 0)
                copy = Board.from_array(board.to_array())
                if not self._solver.solve(copy):
                    board.set_value(row, col, backup)
                attempts -= 1

        for r in range(9):
            for c in range(9):
                board.set_read_only(r, c, board.get_value(r, c) != 0)
